src/shogun/kernel/ExponentialARDKernel.cpp

```diff
@@ -0,0 +1,279 @@
+			set_vector_weights(vec);
+		}
+		else
+			set_scalar_weights(weights[0]);
+	}
+	else
+		set_matrix_weights(weights);
+}
+
+void CExponentialARDKernel::lazy_update_weights()
+{
+	if (parameter_hash_changed())
+	{
+		if (m_ARD_type==KT_SCALAR || m_ARD_type==KT_DIAG)
+		{
+			SGMatrix<float64_t> log_weights(m_log_weights.vector,1,m_log_weights.vlen,false);
+			m_weights_raw=linalg::elementwise_compute(m_log_weights,
+				[  ](float64_t& value)
+				{
+				return CMath::exp(value);
+				});
+		}
+		else if (m_ARD_type==KT_FULL)
+		{
+			m_weights_raw=SGMatrix<float64_t>(m_weights_rows,m_weights_cols);
+			m_weights_raw.set_const(0.0);
+			index_t offset=0;
+			for (int i=0;i<m_weights_raw.num_cols && i<m_weights_raw.num_rows;i++)
+			{
+				float64_t* begin=m_weights_raw.get_column_vector(i);
+				std::copy(m_log_weights.vector+offset,m_log_weights.vector+offset+m_weights_raw.num_rows-i,begin+i);
+				begin[i]=CMath::exp(begin[i]);
+				offset+=m_weights_raw.num_rows-i;
+			}
+		}
+		else
+		{
+			SG_ERROR("Unsupported ARD type\n");
+		}
+		update_parameter_hash();
+	}
+}
+
+SGMatrix<float64_t> CExponentialARDKernel::get_weights()
+{
+	lazy_update_weights();
+	return SGMatrix<float64_t>(m_weights_raw);
+}
+
+void CExponentialARDKernel::set_scalar_weights(float64_t weight)
+{
+	REQUIRE(weight>0, "Scalar (%f) weight should be positive\n",weight);
+	m_log_weights=SGVector<float64_t>(1);
+	m_log_weights.set_const(CMath::log(weight));
+	m_ARD_type=KT_SCALAR;
+
+	m_weights_rows=1.0;
+	m_weights_cols=1.0;
+}
+
+void CExponentialARDKernel::set_vector_weights(SGVector<float64_t> weights)
+{
+	REQUIRE(rhs==NULL && lhs==NULL,
+		"Setting vector weights must be before initialize features\n");
+	REQUIRE(weights.vlen>0, "Vector weight should be non-empty\n");
+	m_log_weights=SGVector<float64_t>(weights.vlen);
+	for(index_t i=0; i<weights.vlen; i++)
+	{
+		REQUIRE(weights[i]>0, "Each entry of vector weight (v[%d]=%f) should be positive\n",
+			i,weights[i]);
+		m_log_weights[i]=CMath::log(weights[i]);
+	}
+	m_ARD_type=KT_DIAG;
+
+	m_weights_rows=1.0;
+	m_weights_cols=weights.vlen;
+}
+
+void CExponentialARDKernel::set_matrix_weights(SGMatrix<float64_t> weights)
+{
+	REQUIRE(rhs==NULL && lhs==NULL,
+		"Setting matrix weights must be before initialize features\n");
+	REQUIRE(weights.num_cols>0, "Matrix weight should be non-empty");
+	REQUIRE(weights.num_rows>=weights.num_cols,
+		"Number of row (%d) must be not less than number of column (%d)",
+		weights.num_rows, weights.num_cols);
+
+	m_weights_rows=weights.num_rows;
+	m_weights_cols=weights.num_cols;
+	m_ARD_type=KT_FULL;
+	index_t len=(2*m_weights_rows+1-m_weights_cols)*m_weights_cols/2;
+	m_log_weights=SGVector<float64_t>(len);
+		
+	index_t offset=0;
+	for (int i=0; i<weights.num_cols && i<weights.num_rows; i++)
+	{
+		float64_t* begin=weights.get_column_vector(i);
+		REQUIRE(begin[i]>0, "The diagonal entry of matrix weight (w(%d,%d)=%f) should be positive\n",
```

Comment:
yes.
For now on, we assume the matrix weight L is a triangular matrix with positive diagonal elements.
The reason is W=L*L'.
* We want W is positive definitive. 
* What is more, if diagonal entries of L are positive, the decomposition is unique, which is good for model selection.